lib/eslint/eslint.js

```diff
@@ -0,0 +1,457 @@
+    }
+
+    return {
+        allowInlineConfig,
+        baseConfig,
+        cache,
+        cacheLocation,
+        configFile,
+        cwd,
+        envs,
+        extensions,
+        fix,
+        fixTypes,
+        globals,
+        globInputPaths,
+        ignore,
+        ignorePath,
+        ignorePattern,
+        parser,
+        parserOptions,
+        plugins: normalizePluginIds(plugins),
+        reportUnusedDisableDirectives,
+        resolvePluginsRelativeTo,
+        rulePaths,
+        rules,
+        useEslintrc
+    };
+}
+
+/**
+ * Create rulesMeta object.
+ * @param {Map<string,Rule>} rules a map of rules from which to generate the object.
+ * @returns {Object} metadata for all enabled rules.
+ */
+function createRulesMeta(rules) {
+    return Array.from(rules).reduce((retVal, [id, rule]) => {
+        retVal.rulesMeta[id] = rule.meta;
+        return retVal;
+    }, { rulesMeta: {} });
+}
+
+/**
+ * Processes the linting results generated by a CLIEngine linting report to
+ * match the ESLint class's API.
+ * @param {LintReport} report The CLIEngine linting report to process.
+ * @returns {LintResult[]} The processed linting results.
+ */
+function processCLIEngineLintReport({ results, usedDeprecatedRules }) {
+    return results.map(result => {
+        result.usedDeprecatedRules = usedDeprecatedRules;
+        return result;
+    });
+}
+
+class ESLint {
+
+    /**
+     * Creates a new instance of the main ESLint API.
+     * @param {ESLintOptions} options The options for this instance.
+     */
+    constructor(options) {
+        const cliEngine = new CLIEngine(processOptions(options));
+
+        if (options.plugins.length) {
+            for (const plugin of options.plugins) {
+                if (isValidPluginObject(plugin)) {
+                    cliEngine.addPlugin(plugin.id, plugin.definition);
+                } else if (typeof plugin !== "string") {
+                    throw new Error("Invalid plugin. Plugins must be specified as a string (e.g., \"eslint-plugin-example\") or as an object (e.g., { id: string; definition: Object }).");
+                }
+            }
+        }
+
+        // Initialize private properties.
+        internalSlotsMap.set(this, {
+            cliEngine
+        });
+    }
+
+    /**
+     * Outputs fixes from the given results to files.
+     * @param {LintReport} report The report object created by CLIEngine.
+     * @returns {Promise<void>} Returns a promise that is used to track side effects.
+     */
+    static async outputFixes(report) {
+        CLIEngine.outputFixes(report);
+    }
+
+    /**
+     * An Array.prototype.sort() compatible compare function to order results by their file path.
+     * @param {LintResult} a The first lint result.
+     * @param {LintResult} b The second lint result.
+     * @returns {number} An integer representing the order in which the two results should occur.
+     */
+    static compareResultsByFilePath(a, b) {
```

Comment:
Picking this up again today, and I'm reconsidering whether this method should be part of this PR. Since we're still running the lint job synchronously, this feels a bit premature, and I think it should be deferred until we actually need it (when parallel linting is implemented). This will allow us to alter the design if we need to and avoid breaking changes. Thoughts?

Reply:
That makes sense. It's unnecessary right now, and leaving it out decreases the surface area of this PR. We're then free to add it in a semver-minor change later.

Reply:
If it’s not needed, good idea to leave it out. It’s a lot easier to add new APIs later than it is to remove them.

Reply:
Thanks for the quick feedback, all. Will defer this addition and remove it from this PR.